src/rng.h

```diff
@@ -94,4 +94,23 @@ inline V random_entry_removed( C &container )
     return result;
 }
 
+namespace cata
+{
+template<typename T>
+class optional;
+} // namespace cata
+
+class map;
+struct tripoint;
+class tripoint_range;
+
+/// Returns a range enclosing all valid points of the map.
+tripoint_range points_in_range( const map &m );
+/// Returns a random point in the given range that satisfies the given predicate ( if any ).
+cata::optional<tripoint> random_point( const tripoint_range &range,
```

Comment:
Do we really need our own implementation of `optional` when we have `invalid_tripoint` for this purpose? Maybe we'd better wait up for the native one.

Reply:
`tripoint_min` is a hack. It's inconsistent (some may use `tripoint_max`, some may only compare the `x` component). `tripoint_min` is a lie - the function does (sometimes) *not* return *any* value at all.

`optional` shows intent - it may or may not return a `tripoint`. This implementation should be compatible with the standard version, so it can be replaced once we can use it.

`optional` allows a nice syntax - `tripoint_min` makes it more complicated.
```C++
if( auto result = function( ... ) ) {
	auto value = *result;
	// use value
} else {
	// handle non-existing value
}
```